File: ios/Pitwall/Services/APIClient.swift
import Foundation

// MARK: - Rate limiter (actor for thread safety)

actor RateLimiter {
    private var tokens: Double
    private let maxTokens: Double
    private let refillRate: Double  // tokens per second
    private var lastRefill: Date = .now

    init(maxTokens: Double, refillRate: Double) {
        self.maxTokens = maxTokens
        self.refillRate = refillRate
        self.tokens = maxTokens
    }

    func acquire() async {
        refill()
        if tokens >= 1 {
            tokens -= 1
            return
        }
        let waitSeconds = 1.0 / refillRate
        let waitNanos = UInt64(waitSeconds * 1_000_000_000)
        try? await Task.sleep(nanoseconds: waitNanos)
        refill()
        tokens -= 1
    }

    private func refill() {
        let now = Date()
        let elapsed = now.timeIntervalSince(lastRefill)
        tokens = min(maxTokens, tokens + elapsed * refillRate)
        lastRefill = now
    }
}

// MARK: - Shared limiters

let openF1Limiter = RateLimiter(maxTokens: 3, refillRate: 3)
let jolpicaLimiter = RateLimiter(maxTokens: 4, refillRate: 4)

// MARK: - API error

enum APIError: LocalizedError {
    case httpError(Int, String)
    case decodingError(Error)
    case invalidURL

    var errorDescription: String? {
        switch self {
        case .httpError(let code, let url): return "HTTP \(code) for \(url)"
        case .decodingError(let e): return "Decoding error: \(e.localizedDescription)"
        case .invalidURL: return "Invalid URL"
        }
    }
}

// MARK: - Fetch helpers

func fetchOpenF1<T: Decodable>(_ endpoint: String, params: [String: String] = [:]) async throws -> [T] {
    await openF1Limiter.acquire()
    var components = URLComponents(string: "https://api.openf1.org/v1/\(endpoint)")!
    components.queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) }
    guard let url = components.url else { throw APIError.invalidURL }
    return try await fetchJSON([T].self, from: url)
}

func fetchJolpica<T: Decodable>(_ path: String) async throws -> T {
    await jolpicaLimiter.acquire()
    guard let url = URL(string: "https://api.jolpi.ca/ergast/f1/\(path)") else {
        throw APIError.invalidURL
    }
    return try await fetchJSON(T.self, from: url)
}

private func fetchJSON<T: Decodable>(_ type: T.Type, from url: URL) async throws -> T {
    let (data, response) = try await URLSession.shared.data(from: url)
    if let http = response as? HTTPURLResponse, http.statusCode != 200 {
        throw APIError.httpError(http.statusCode, url.absoluteString)
    }
    do {
        return try JSONDecoder().decode(type, from: data)
    } catch {
        throw APIError.decodingError(error)
    }
}

File: web/src/services/api-client.ts
export class ApiError extends Error {
  constructor(
    message: string,
    public status: number,
    public url: string,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

class RateLimiter {
  private tokens: number;
  private lastRefill: number;

  constructor(
    private maxTokens: number,
    private refillRate: number,
  ) {
    this.tokens = maxTokens;
    this.lastRefill = Date.now();
  }

  async acquire(): Promise<void> {
    this.refill();
    if (this.tokens > 0) {
      this.tokens--;
      return;
    }
    const waitMs = (1 / this.refillRate) * 1000;
    await new Promise((resolve) => setTimeout(resolve, waitMs));
    this.refill();
    this.tokens--;
  }

  private refill(): void {
    const now = Date.now();
    const elapsed = (now - this.lastRefill) / 1000;
    this.tokens = Math.min(this.maxTokens, this.tokens + elapsed * this.refillRate);
    this.lastRefill = now;
  }
}

const openf1Limiter = new RateLimiter(3, 3);
const jolpicaLimiter = new RateLimiter(4, 4);

export async function fetchOpenF1<T>(
  endpoint: string,
  params?: Record<string, string | number>,
): Promise<T[]> {
  await openf1Limiter.acquire();
  const url = buildUrl("https://api.openf1.org/v1", endpoint, params);
  return fetchJson<T[]>(url);
}

export async function fetchJolpica<T>(path: string): Promise<T> {
  await jolpicaLimiter.acquire();
  const url = `https://api.jolpi.ca/ergast/f1/${path}`;
  return fetchJson<T>(url);
}

function buildUrl(
  base: string,
  endpoint: string,
  params?: Record<string, string | number>,
): string {
  const url = new URL(`${base}/${endpoint}`);
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      url.searchParams.set(key, String(value));
    }
  }
  return url.toString();
}

async function fetchJson<T>(url: string): Promise<T> {
  const response = await fetch(url);
  if (!response.ok) {
    throw new ApiError(
      `API request failed: ${response.status} ${response.statusText}`,
      response.status,
      url,
    );
  }
  return response.json() as Promise<T>;
}
